pubtools/pulplib/_impl/client/client.py

```diff
@@ -179,6 +182,28 @@ def search_repository(self, criteria=None):
             response_f, lambda data: self._handle_page(Repository, search, data)
         )
 
+    def _do_upload_file(self, upload_id, file_obj):
+        def do_next_upload(checksum, size):
+            data = file_obj.read(self._CHUNK_SIZE)
+            if data:
+                checksum.update(data)
+                return f_flat_map(
+                    self._do_upload(data, upload_id, size),
+                    lambda _: do_next_upload(checksum, size + len(data)),
+                )
+            else:
+                # nothing more to upload, return checksum and size
+                if is_string:
+                    file_obj.close()
+                return f_return((checksum.hexdigest(), size))
+
+        is_string = isinstance(file_obj, six.string_types)
+        if is_string:
+            file_obj = open(file_obj, "rb")
+
+        LOG.info("Uploading %s to Pulp", file_obj)
+        return f_flat_map(f_return(), lambda _: do_next_upload(hashlib.sha256(), 0))
```

Comment:
I was thinking at first that this was pointless, then I realized it's actually not because it means that if do_next_upload fails, that will result in a future which fails rather than this method raising an exception directly.

It could be considered desirable because it means there's just one method for closing file_obj in both cases (see earlier comment).

Reply:
I guess this means good?